### ISSUE TITLE
undefined symbols in libopenvdb_houdini

### ISSUE THREAD

**Issue by alexwheezy** [AcademySoftwareFoundation/openvdb#343]
After compiling the library from source with all paths and dependent libraries, I still get an error about an undefined symbol. I tried to collect on Debian 9.7 (Stretch).

I tried to specify the path to load the new OpenVDB library in export LD_LIBRAR_PATH=/home/alex/lib/openvdb/6.0.1/lib:$LD_LIBRARY_PATH, but this does not help.

`readelf -d libopenvdb_houdini.so`

```
Dynamic section at offset 0x2557e8 contains 41 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libGL.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libX11.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libXext.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libXi.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libHalf.so.12]
 0x0000000000000001 (NEEDED)             Shared library: [libtbb.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_iostreams.so.1.62.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_system.so.1.62.0]
 0x0000000000000001 (NEEDED)             Shared library: [libopenvdb.so.6.0]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000001d (RUNPATH)            Library runpath: [/opt/hfs17.0.479/dsolib:/home/alex/Libs/openvdb/6.0.1/lib]
 0x000000000000000c (INIT)               0xc7c98
 0x000000000000000d (FINI)               0x207d44
 0x0000000000000019 (INIT_ARRAY)         0x44dc00
 0x000000000000001b (INIT_ARRAYSZ)       80 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x44dc50
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x1f0
 0x0000000000000005 (STRTAB)             0x22490
 0x0000000000000006 (SYMTAB)             0x6d10
 0x000000000000000a (STRSZ)              532913 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x457000
 0x0000000000000002 (PLTRELSZ)           32544 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0xbfd78
 0x0000000000000007 (RELA)               0xa6bd8
 0x0000000000000008 (RELASZ)             102816 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffe (VERNEED)            0xa6ae8
 0x000000006fffffff (VERNEEDNUM)         5
 0x000000006ffffff0 (VERSYM)             0xa4642
 0x000000006ffffff9 (RELACOUNT)          320
 0x0000000000000000 (NULL)               0x0
```

`nm -u libopenvdb_houdini.so`

```
                 U _ZNK7OP_Node17getParallelInputsER10OP_ContextR11UT_ValArrayIPS_E
                 U _ZNK7OP_Node17getPathWithSubnetER9UT_String
                 U _ZNK7OP_Node18canCreateNewOpTypeEv
                 U _ZNK7OP_Node18getDescriptiveNameER9UT_String
                 U _ZNK7OP_Node18getSaveFlagsStringER9UT_StringPKcb
                 U _ZNK7OP_Node19getNumVisibleInputsEv
                 U _ZNK7OP_Node19matchChildBackwardsEPKcPS_S1_i
                 U _ZNK7OP_Node20createParmMicroNodesER8PRM_Parm
                 U _ZNK7OP_Node20getDataFileExtensionEi
                 U _ZNK7OP_Node20getNumVisibleOutputsEv
                 U _ZNK7OP_Node21isParmPendingOverrideEPKci
                 U _ZNK7OP_Node21pushAsPwdAndRunPythonER15PY_CompiledCodeN9PY_Result4TypeERS2_P20PY_EvaluationContext
                 U _ZNK7OP_Node22getDescriptiveParmNameER9UT_String
                 U _ZNK7OP_Node22getInputReferenceConstEj
                 U _ZNK7OP_Node23hasDifferentMultiInputsEv
                 U _ZNK7OP_Node25evaluateDisableExpressionERK8PRM_ParmRK12UT_StringRef
                 U _ZNK7OP_Node25getPreParmSaveFlagsStringER9UT_String
                 U _ZNK7OP_Node26convertOpdefToAbsolutePathER9UT_String
                 U _ZNK7OP_Node26getPostParmSaveFlagsStringER9UT_String
                 U _ZNK7OP_Node28getExtraNodesForCopyOrDeleteER11UT_ValArrayIPS_E
                 U _ZNK7OP_Node28isInputConnectedInsideSubnetEi
                 U _ZNK7OP_Node33isNamedInputConnectedInsideSubnetERK14OP_ConnectorId
                 U _ZNK7OP_Node5isVexEv
                 U _ZNK7OP_Node7getFlagEc
                 U _ZNK7OP_Node7nInputsEv
                 U _ZNK7OP_Node8getColorEv
                 U _ZNK7OP_Node8getInputEjb
                 U _ZNK7OP_Node9getUnloadEv
                 U _ZNK7OP_Node9isManagerEv
                 U _ZNK7OP_Node9maxInputsEv
                 U _ZNK7OP_Node9minInputsEv
                 U _ZNK8GA_Range5beginEv
                 U _ZNK8PRM_Parm16isFactoryDefaultEv
                 U _ZNK8SOP_Node11getOpTypeIDEv
                 U _ZNK8SOP_Node12getChildTypeEv
                 U _ZNK8SOP_Node14getChildTypeIDEv
                 U _ZNK8SOP_Node14getMemoryUsageEb
                 U _ZNK8SOP_Node14inputGeoHandleEi
                 U _ZNK8SOP_Node16getFileExtensionEi
                 U _ZNK8SOP_Node17getCookedDataTypeEv
                 U _ZNK8SOP_Node20getLoadedMemoryUsageEv
                 U _ZNK8SOP_Node36getDefaultSelectedInfoTreeBranchNameEv
```
`ldd -d -r libopenvd_houdini.so`

```
undefined symbol: _ZNK16GA_PrimitiveList16hardenPrimitivesEv	(./libopenvdb_houdini.so)
undefined symbol: _ZN17GT_GEOPrimCollectC2Ev	(./libopenvdb_houdini.so)
undefined symbol: _ZNK15UT_BoundingBoxTIfE13getBBoxPointsERA8_11UT_Vector3TIfE	(./libopenvdb_houdini.so)
undefined symbol: _ZN12PRM_TemplateC1E8PRM_TypeiP8PRM_NameP11PRM_DefaultP14PRM_ChoiceListP9PRM_Range12PRM_CallbackP13PRM_SpareDataiPKcP19PRM_ConditionalBase	(./libopenvdb_houdini.so)
undefined symbol: _ZNK15GA_DataBitArray3getEl	(./libopenvdb_houdini.so)
undefined symbol: _ZN16GT_GEOOffsetList6concatERKS_	(./libopenvdb_houdini.so)
undefined symbol: _ZN12UT_Interrupt11opInterruptEi	(./libopenvdb_houdini.so)
undefined symbol: _ZN12GT_DataArraydlEPvm	(./libopenvdb_houdini.so)
undefined symbol: _ZN7openvdb4v6_08GridBase7setNameERKSs	(./libopenvdb_houdini.so)
undefined symbol: _ZN15GT_AttributeMapD1Ev	(./libopenvdb_houdini.so)
undefined symbol: _ZN9GA_Detail16destroyPrimitiveER12GA_Primitiveb	(./libopenvdb_houdini.so)
undefined symbol: _ZN11OP_Operator11setIconNameEPKc	(./libopenvdb_houdini.so)
undefined symbol: _ZN7openvdb4v6_06points26StringAttributeWriteHandle3setEjjRKSs	(./libopenvdb_houdini.so)
undefined symbol: _ZN10GEO_Detail14findFloatTupleE17GA_AttributeOwner17GA_AttributeScopeRK12UT_StringRefii	(./libopenvdb_houdini.so)
undefined symbol: _ZN9UT_Thread16getNumProcessorsEv	(./libopenvdb_houdini.so)
undefined symbol: _ZN8SOP_NodeC2EP10OP_NetworkPKcP11OP_Operator	(./libopenvdb_houdini.so)
undefined symbol: _ZN10GU_PrimVDB34createMetadataFromGridAttrsAdapterEPvRK11GEO_PrimVDBRK10GEO_Detail	(./libopenvdb_houdini.so)
undefined symbol: _ZN10OBJ_Camera24addInterestOnCameraParmsEP7OP_Node	(./libopenvdb_houdini.so)
undefined symbol: _ZN11GA_IteratorC1ERKS_	(./libopenvdb_houdini.so)
undefined symbol: _ZNK9UT_String11toStdStringEv	(./libopenvdb_houdini.so)
undefined symbol: _ZNK12GT_DataArray10getU8ArrayER15UT_IntrusivePtrIS_E	(./libopenvdb_houdini.so)
undefined symbol: _ZNK12GT_DataArray14t_extendedFillIfEEvPT_lliii	(./libopenvdb_houdini.so)
undefined symbol: _ZNK7openvdb4v6_06points12AttributeSet10Descriptor4typeEm	(./libopenvdb_houdini.so)
undefined symbol: _ZN12GEO_PrimPoly5buildEP9GA_Detaillbb	(./libopenvdb_houdini.so)
undefined symbol: _ZN10UT_IStream5breadEPcl	(./libopenvdb_houdini.so)
undefined symbol: _ZNK12GT_DataArray14t_extendedFillIsEEvPT_lliii	(./libopenvdb_houdini.so)
undefined symbol: _ZN15GU_DetailHandle9writeLockEv	(./libopenvdb_houdini.so)
undefined symbol: _ZN17GA_RWAttributeRefC1EP12GA_Attribute	(./libopenvdb_houdini.so)
undefined symbol: _ZN9UT_StringaSERK15UT_StringHolder	(./libopenvdb_houdini.so)
undefined symbol: _ZN16GT_PrimCurveMesh4initE8GT_BasisRK15UT_IntrusivePtrI12GT_DataArrayERKS1_I16GT_AttributeListES9_S9_b15GT_IndexingMode	(./libopenvdb_houdini.so)
undefined symbol: _ZN12PRM_TemplateaSERKS_	(./libopenvdb_houdini.so)
undefined symbol: _ZN7openvdb4v6_06points12AttributeSet10Descriptor6createERKSt4pairISsSsE	(./libopenvdb_houdini.so)
undefined symbol: _ZN8SOP_Node18fillAttribNameMenuEP8PRM_Namei17GA_AttributeOwneriPFbPK12GA_AttributePvES6_	(./libopenvdb_houdini.so)
undefined symbol: _ZN13GA_GroupTable8newGroupERK15UT_StringHolder	(./libopenvdb_houdini.so)
undefined symbol: _ZN12GA_ATIString9setStringElRK15UT_StringHolderi	(./libopenvdb_houdini.so)
undefined symbol: _ZN17GA_PrimitiveGroup3addEPK12GA_Primitive	(./libopenvdb_houdini.so)
undefined symbol: _ZN8SOP_Node24fillInfoTreeNodeSpecificER11UT_InfoTreeRK20OP_NodeInfoTreeParms	(./libopenvdb_houdini.so)
undefined symbol: _ZN7openvdb4v6_08Metadata14createMetadataERKSs	(./libopenvdb_houdini.so)
undefined symbol: _ZN23theTmpDirLockStaticInitC1Ev	(./libopenvdb_houdini.so)
undefined symbol: _ZN12SOP_NodeVerb10lookupVerbERK12UT_StringRef	(./libopenvdb_houdini.so)
undefined symbol: _ZN8SOP_NodeD2Ev	(./libopenvdb_houdini.so)
undefined symbol: _ZN9UT_Thread18isThreadingEnabledEv	(./libopenvdb_houdini.so)
undefined symbol: _ZN8SOP_Node15duplicateSourceEjR10OP_ContextP9GU_Detailb	(./libopenvdb_houdini.so)
undefined symbol: _ZN12GT_DataArrayC2Ev	(./libopenvdb_houdini.so)
undefined symbol: _ZN22GA_GBPrimitiveIterator7advanceEv	(./libopenvdb_houdini.so)
undefined symbol: _ZN10GEO_DetailC2ER19GA_PrimitiveFactoryb	(./libopenvdb_houdini.so)
undefined symbol: _ZN7OP_Node9setExposeEi	(./libopenvdb_houdini.so)
undefined symbol: _ZN12GT_DataArraynwEm	(./libopenvdb_houdini.so)
undefined symbol: _ZNK7openvdb4v6_06points12AttributeSet10groupIndexERKSs	(./libopenvdb_houdini.so)
```
Any ideas on how to fix this?

**Comment by Idclip** [comment]
Hi @alexwheezy - would you be able to try with the latest verison of VDB (either 6.1 or the current master branch) and let me know if you still see the issue?

Thanks

**Comment by alexwheezy** [comment]
> Hi @alexwheezy - would you be able to try with the latest verison of VDB (either 6.1 or the current master branch) and let me know if you still see the issue?
> 
> Thanks

Hi, @Idclip

Yes, I already tried to download version 6.1, but there was a similar error until I added add_definitions (-D_GLIBCXX_USE_CXX11_ABI = 0) to the root CMakeLists.

And it helped solve my problem.

**Comment by Idclip** [comment]
Thanks @alexwheezy, glad to hear that solved your problem! We'll look to getting that definition added, though I believe it's only applicable for certain versions of Houdini. Can I double check what version of Houdini you were build for? I'm guessing 16.5?

**Comment by alexwheezy** [comment]
@Idclip, I tried to build a project under H17.0 and H17.5, while everything works at the moment.

**Comment by Idclip** [comment]
Thanks for the info @alexwheezy - I have a PR here which should fix this going forward:

https://github.com/AcademySoftwareFoundation/openvdb/pull/506

Keeping this issue open until it's merged

**Comment by Idclip** [comment]
This fix has been merged and released in 6.2.0. Please feel free to re-open if you continue to experience problems!

**Comment by alexwheezy** [comment]
Wow! Thank you very much, @Idclip! I will certainly test this update.
